std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,247 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Finds the mean (colloquially known as the average) of a range.
+This function is $(BIGOH r.length).
+
+Params:
+    r = an input range
+
+Returns:
+    the mean of r
+*/
+auto mean(R)(R r)
+if (isInputRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+}
+body
+{
+    static if (hasLength!R)
+    {
+        if (r.length == 0)
+        {
+            return double.nan;
+        }
+        else if (r.length == 1)
+        {
+            return r.front;
+        }
+
+        return sum(r) / r.length;
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
+            (tuple(size_t(0), 0.0), r);
+
+
+        if (pair[0] == 0)
+        {
+            return double.nan;
+        }
+
+        return pair[1] / pair[0];
+    }
+}
+
+///
+@safe pure nothrow unittest
+{
+    import std.math : approxEqual;
+
+    assert(mean([1, 2, 3]) == 2);
+    assert(approxEqual(mean([1.5, 2.5, 12.5]), 5.5));
+}
+
+@safe pure nothrow unittest
+{
+    import std.internal.test.dummyrange : ReferenceInputRange;
+    import std.math : approxEqual;
+
+    auto r1 = new ReferenceInputRange!int([1, 2, 3]);
+    assert(r1.mean == 2);
+
+    auto r2 = new ReferenceInputRange!double([1.5, 2.5, 12.5]);
+    assert(approxEqual(r2.mean, 5.5));
+}
+
+/**
+Function to compute the quantile of a $(D SortedRange) given $(D q), which
+is a number between 0 and 1. If the range is empty, this function will fail
+an assert in debug mode and return NaN in release mode. Is $(BIGOH r.length * q)
+for forward ranges and $(BIGOH 1) for random access ranges.
+
+Params:
+    r = a forward range with length or a random access range
+    q = where to calculate the quantile
+
+Returns:
+    the quantile of the range given $(D q)
+*/
+double quantile(alias less = "a < b", R)(SortedRange!(R, less) r, double q)
+    if (isForwardRange!R &&
+        hasLength!R &&
+        !isInfinite!R &&
+        is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+    assert(q > 0 && q < 1);
+}
+body
+{
+    import std.math : modf, approxEqual;
+    import std.algorithm.searching : countUntil;
+
+    if (r.length == 0)
+    {
+        return double.nan;
+    }
+    else if (r.length == 1)
+    {
+        return r.front;
+    }
+
+    if (less.countUntil(">") > -1)
```

Comment:
Yeah, this line is extremely scary, because a user-provided `less` predicate, even if it's a string, may not have any `<` or `>` symbol in it. E.g., if it's a custom float type or some such, and the predicate calls an internal function for comparisons, e.g., `a.myFloatType.util.isLessThan(b)`.